#166840923 Feature - User should be able rate articles

codexempire opened the pull request:
### What does this PR do?
Enables users to rate an article

### Description of Tasks to be completed?
- enabled users the ability to rate an article
How should this be manually tested?
- Clone the repo, cd into it and run npm run start:dev
- make request to 'localhost:3000/api/v1/articles/:articleId/:ratings'
**NB**:  **articleId** refers to the article's id to be rated while **ratings** are a number between 1and 5

Any background context you want to provide?

What are the relevant pivotal tracker stories?
#166840923

tolumide-ng commented on .travis.yml:
```diff
@@ -12,6 +12,7 @@ env:
   global:
     - NODE_ENV=test
     - JWT_SECRET=prfhdfdh
+    - TEST_TOKEN=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MiwidXNlcm5hbWUiOiJUaW1pVGVqdW1vbGEiLCJlbWFpbCI6InRpbWl0ZWp1bW9sYUBhbmRlbGEuY29tIiwicm9sZSI6InVzZXIiLCJpYXQiOjE1NjI2NzY2Nzl9.AFg3xbbcm_hWBB8JjyXdoHIk-mG-jpq2gZAgUU9-tl8
```
Hello @codexempire , could you please remove this?
I do not  think you should have a token here

tolumide-ng commented on src/controllers/articles.js:
```diff
@@ -0,0 +1,96 @@
+      ...req.body
+    })
+      .then(article =>
+        res.status(201).send({
```
Good work @codexempire , could you check the develop branch and use the response generator class to generate your responses. Kindly reuse this in other parts of your code.

tolumide-ng commented on src/controllers/articles.js:
```diff
@@ -0,0 +1,96 @@
+  },
+
+  retrieve(req, res) {
+    return Article.findByPk(req.params.articleId, {})
```
Please create this method in your baserepository if it does not already exist

codexempire replied:
Hi @tolumide-ng the implementation of the article is just a dummy for my feature to be able to run

tolumide-ng replied:
Oh! Okay, could you ensure you remove this before it is merged?

segunolalive replied:
Remove the faux article controller. You can seed the `db` with articles instead. That way, you can test your code.

tolumide-ng commented on src/database/models/ratings.js:
```diff
@@ -0,0 +1,43 @@
+    {}
+  );
+  Ratings.associate = models => {
+    // associations can be defined here
```
can you remove the comments pls

tolumide-ng commented on src/test/ratings.spec.js:
```diff
@@ -0,0 +1,54 @@
+chai.use(chaiHttp);
+chai.should();
+
+describe('Testing Rating of articles', () => {
```
could you restructure your tests so that it truncates the table after each test to avoid dependencies

tolumide-ng reviewed (commented):
Great work @codexempire , kindly implement all the changes recommended

tolumide-ng commented on package.json:
```diff
@@ -86,6 +86,8 @@
         "prettier-eslint": "^9.0.0",
         "prettier-eslint-cli": "^5.0.0",
-        "sinon": "^7.3.2"
+        "rewire": "^4.0.1",
```
This dependency seems to be out-of-date according to the documentation on [npm](https://www.npmjs.com/package/rewire), Please remove it. Also why do you have `slug` as a dependency?

tolumide-ng commented on src/controllers/articles.js:
```diff
@@ -0,0 +1,96 @@
+      ...req.body
+    })
+      .then(article =>
+        res.status(201).send({
```
Please change this to the default responseGenerator: Please check here for more information

tolumide-ng commented on src/controllers/articles.js:
```diff
@@ -0,0 +1,96 @@
+    return Article.findByPk(req.params.articleId, {})
+      .then(article => {
+        if (!article) {
+          return res.status(404).send({
```
Please use the default response generator

tolumide-ng commented on src/controllers/articles.js:
```diff
@@ -0,0 +1,96 @@
+        }
+        return article
+          .update({
+            title: req.body.title || article.title,
```
Nice implementation but I think you could destructure the `article` and the `req.body` first to make your code cleaner

tolumide-ng commented on src/database/migrations/20190703214918-create-article.js:
```diff
@@ -0,0 +1,44 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => {
```
I notice you already have three migrations for the `Articles` model, is there are reason why?

tolumide-ng commented on src/database/migrations/20190704115631-create-users.js:
```diff
@@ -0,0 +1,32 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => {
```
Nice work, Is there a reason why you are running this new migrations on the `users` table? I do not see a difference between this and the current one on the repository.

tolumide-ng commented on src/database/seeders/20190704205541-demo-articles.js:
```diff
@@ -0,0 +1,40 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.bulkInsert(
```
Could you use `faker` to create the seeders?

tolumide-ng commented on src/middleware/validateRating.js:
```diff
@@ -0,0 +1,32 @@
+import Joi from 'joi';
+
+export default {
```
Nice implementation, could you check the way the validations was implemented in the [validationSchema.js](https://github.com/andela/ah-kifaru-backend/blob/develop/src/middleware/validation.schema.js) module?, I do not think you need another module for this. I think this would help us maintain consistency in our codebase

tolumide-ng commented on src/repository/base.repository.js:
```diff
@@ -64,6 +64,17 @@ class BaseRepository {
+   * @returns {object} - returns an database object
+   * @memberof BaseRepository
+   */
+  static async updateField(data, options) {
```
Please check the current baseRepository, there is already a method for this in the repository

codexempire replied:
Yes I did notice but I added this because the `update` is not actually working well for the implementation of my feature

tolumide-ng commented on src/routes/v1/index.js:
```diff
@@ -1,8 +1,10 @@
   app.use('/auth', authRoutes);
   app.use('/api/v1/users', userRoutes);
   app.use('/api/v1/auth', userRoutes);
+  app.use('/api/v1', ratingArticleRoutes);
```
I think there could be a better way to implement this route. You could leave it on the `users` route since it is tied to the user

codexempire replied:
The route is tied to the `articles` and cannot be kept within the `user` file. It is related to both the `articles` and `user`. The route is `/api/v1/articles/:articleId/:ratings`.

segunolalive replied:
why are you mounting the ratings controller on the base route. Follow the example from others. As a matter if fact, rating article should be a method on the `ArticleController`. Therefore the route should be in the article routes. Communicate with @tobechukwuobitube to handle that

tolumide-ng replied:
In that case, I think it should be `articles` here

tolumide-ng commented on src/test/ratings.spec.js:
```diff
@@ -0,0 +1,51 @@
+chai.use(chaiHttp);
+chai.should();
+
+describe('Testing Rating of articles', () => {
```
Please apply the TTL's corrections in the implementation of your `description` and `it`. 
e.g. ```describe('HTTP-METHOD, route_path', () => {})```

codexempire replied:
okay thanks, I will.

tolumide-ng reviewed (changes requested):
Great work @codexempire , please implement all the recommended changes

tolumide-ng commented on package.json:
```diff
@@ -86,6 +86,8 @@
         "prettier": "^1.18.2",
         "prettier-eslint": "^9.0.0",
         "prettier-eslint-cli": "^5.0.0",
-        "sinon": "^7.3.2"
```
Please return this dependency as it is used in other parts of the code

codexempire replied:
Hi @tolumide-ng thanks for your observation but just checked the `package.json` of the `develop branch` and found the dependency version to be what is available on the `develop branch`.

segunolalive commented on src/helpers/ratingsInsertion.js:
```diff
@@ -0,0 +1,41 @@
+import db from '../database/models';
+import BaseRepository from '../repository/base.repository';
```
move this logic to your controller.

This is a controller.

segunolalive replied:
helpers should be small utility functions that are used across controller methods

segunolalive commented on src/middleware/validateRating.js:
```diff
@@ -0,0 +1,32 @@
+    );
+    if (error) {
+      const errs = error.details[0].message.replace(/['"]/g, '');
+      return res.status(400).json({
```
all responses should use the response generator

segunolalive commented on src/test/ratings.spec.js:
```diff
@@ -0,0 +1,51 @@
+chai.should();
+
+describe('Testing Rating of articles', () => {
+  it('it should return a 400 status code if the request is bad', done => {
```
what makes the request bad?

codexempire replied:
That is when the values inputted are failed the validations and are not numbers

segunolalive commented on src/test/ratings.spec.js:
```diff
@@ -0,0 +1,51 @@
+      .request(app)
+      .patch(`/api/v1/articles/:articles/:ratings`)
+      .end((error, response) => {
+        response.body.message.should.be.a('string');
```
This test doesn't demonstrate that a relevant error message was sent. An empty string is a string. What is the message. We should know the exact message that was sent.

segunolalive commented on src/test/ratings.spec.js:
```diff
@@ -0,0 +1,51 @@
+      .patch(`/api/v1/articles/${10}/${4}`)
+      .set('x-access-token', process.env.TEST_TOKEN)
+      .end((error, response) => {
+        expect(response.body.data).to.have.equal('Article does not exist');
```
This is good. You tested the exact message. However, change `data` to `message`

segunolalive commented on src/test/ratings.spec.js:
```diff
@@ -0,0 +1,51 @@
+  it('Should return message if the article to be rated was not found', done => {
+    chai
+      .request(app)
+      .patch(`/api/v1/articles/${10}/${4}`)
```
the rating should be the request body and the route should be `/articles/:id/ratings`

segunolalive commented on src/test/ratings.spec.js:
```diff
@@ -0,0 +1,51 @@
+        expect(response.status).to.equal(200);
+        expect(response.body.data.ratings).to.equal('4');
+        expect(response.body.data.articleId).to.equal(1);
+        expect(response.body.data.userId).to.equal(2);
```
Why are we returning the `userId` after rating? What `userId` is this? The person who rated or the author of the article?
In either case, it provides no added value.

segunolalive commented on src/test/ratings.spec.js:
```diff
@@ -0,0 +1,51 @@
+      .set('x-access-token', process.env.TEST_TOKEN)
+      .end((error, response) => {
+        expect(response.status).to.equal(200);
+        expect(response.body.data.ratings).to.equal('4');
```
These numbers seem like magic numbers

segunolalive reviewed (changes requested):
Looking good. Attend to the comments.

codexempire committed:
Feature(rateArticles) Rating Articles

made middleware for the checks
moved all implementation to the controllers
changed route naming
[Delivers #166840923]

codexempire commented:
Hi found the branch was not accepting my pushes raised the PR with another branch with related naming
